### PR TITLE
[DOC] Format specs

### PR DESCRIPTION
**Pull request by BurdetteLamar**
This new document would eventually replace the format spec discussion at Kernel#sprintf, which would then link to the new doc.

One of the aims here is to make the discussion of a type specifier, say ```%b```, sufficient in itself, without the reader having to carry a lot of info from other sections.  In this beginning, only the one type specifier is detailed; others will follow in new PRs.

### CODE REVIEW

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:16]: Only `Kernel#sprintf` returns the string, the other methods print the string to the related IO/stdout, then return nil.

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:148]: `s/when +0b+ or +0B+ is prefixed/when +#+ flag is used/`

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:184]: Incorrect, it still left pads with 0:

```ruby
sprintf("%020B", 100)
# => "00000000000001100100"
sprintf("%20B", 100)
# => "             1100100"
```

**Review by jeremyevans — CHANGES_REQUESTED** [review]
See inline responses.  Note that pushing additional commits after asking for a review can make review more difficult.

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc:186]: As I said, "One of the aims here is to make the discussion of a type specifier, say %b, sufficient in itself, without the reader having to carry a lot of info from other sections."  A reader trying to look up %b behavior should not have to read the whole page to learn about flags, width, and precision.

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc:142]: As I said, "In this beginning, only the one type specifier is detailed; others will follow in new PRs."  A full-treatment PR would be too huge to review?  Also, I'd like to get agreed on the principles before doing more.

**Comment by BurdetteLamar** [comment]
> Note that pushing additional commits after asking for a review can make review more difficult.

Sorry.  I failed to resist the temptation to plug holes and correct errors.  Will do better.

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [reply doc/kernel/format_specifications.rdoc:186]: Duplicating repetitive content in every section is much worse than documenting shared behavior in one place, and specific exceptions for each specifier, IMO.  If you want to have a bunch of long sections that mostly repeat the same information for different types, you'll have to find another reviewer, because I cannot approve that approach.

Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc:186]: > Duplicating repetitive content in every section is much worse than documenting shared behavior in one place, and specific exceptions for each specifier, IMO. If you want to have a bunch of long sections that mostly repeat the same information for different types, you'll have to find another reviewer, because I cannot approve that approach.

Ditch the repetitive text, but keep the examples?

Comment (jeremyevans) [reply doc/kernel/format_specifications.rdoc:186]: Have a single example section for each type specifier section showing basic behavior without flags. Beyond that, only keep examples for flags that show type specifier-specific behavior, and those should have text explaining the different behavior.  As I mentioned, for `b` and `B`, the `..` handling for negative numbers and `#` for changing capitalization are exceptions that should be mentioned.  There may be other exceptions, but I think only exceptions should have documentation and examples.  Any flag behavior that is the same across type specifiers does not need explanations or examples in each type specifier section.

Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc:186]: @jeremyevans, we're very far apart on how to approach this topic.  With your permission, I'll close this PR.

Comment (jeremyevans) [reply doc/kernel/format_specifications.rdoc:186]: Sure, that's fine.

**Comment by BurdetteLamar** [comment]
I'm re-opening this because I really really want to see improvement to the doc for format specs.

I'll correct the errors cited by @jeremyevans, and seek common ground about the way the types should be documented.

**Comment by BurdetteLamar** [comment]
@jeremyevans, let's try for common ground on ```b``` and ```B```.

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:174]: This example is wrong, should be `sprintf('%#B', 1)  # => "0B1"`

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:177]: No reason to mention modifiers that are the same for all type specifiers.  We don't need the same redundant links in every type specifier section.

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:188]: Incorrect:

```ruby
sprintf('%-14B', 1)
# => "1             "
sprintf('%14B', 1)
# => "             1"
```

File: doc/kernel/format_specifications.rdoc
Comment (jeremyevans) [inline doc/kernel/format_specifications.rdoc:192]: Doesn't appear to be disallowed:

```ruby
sprintf('%14.5b', 3)
# => "         00011"
```

**Review by jeremyevans — CHANGES_REQUESTED** [review]

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc]: Fixed.

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc]: Removed.

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc]: Removed.

File: doc/kernel/format_specifications.rdoc
Comment (BurdetteLamar) [reply doc/kernel/format_specifications.rdoc]: Removed.

**Comment by BurdetteLamar** [comment]
@jeremyevans, I'd originally thought to include this file in kernel.c, but later realized that it'll be better as a free-standing doc that all 4 *printf methods can link to.  Therefore should be in doc/ not doc/kernel?  If so, should the move be before this merge, or after?

**Comment by BurdetteLamar** [comment]
@jeremyevans Oh, and what about the many to-be-supplied sections?  Keep adding to this PR, reviewing as we go?  Or merge and add later?

**Comment by jeremyevans** [comment]
> @jeremyevans, I'd originally thought to include this file in kernel.c, but later realized that it'll be better as a free-standing doc that all 4 *printf methods can link to. Therefore should be in doc/ not doc/kernel? If so, should the move be before this merge, or after?

Let's move to `doc/` before the merge.

> @jeremyevans Oh, and what about the many to-be-supplied sections? Keep adding to this PR, reviewing as we go? Or merge and add later?

Let's merge this after the move to `doc/`, then you can work on other specifiers.

**Comment by BurdetteLamar** [comment]
@jeremyevans, all set?  I've removed the addition (kernel) to doc/.document.

**Review by jeremyevans — APPROVED** [review]

**Pull request merged**
